classes/question_bridge_base.php

```diff
@@ -0,0 +1,271 @@
+     *
+     * @param \stdClass $userdata
+     * @param array $userfieldsmapping as returned by {@see self::get_user_fields_api_mapping()}
+     * @return array
+     */
+    protected function extract_user_attributes(\stdClass $userdata, array $userfieldsmapping): array {
+        $attributes = [];
+
+        foreach ($userfieldsmapping as $attributename => $field) {
+            if ($attributename === 'display_name') {
+                $attributes[$attributename] = \core_user::get_fullname($userdata, $this->context);
+            } else {
+                $attributes[$attributename] = $userdata->{$field} ?? null;
+            }
+        }
+
+        return $attributes;
+    }
+
+    /**
+     * Get the attributes of a group and its members.
+     *
+     * @param int $groupid
+     * @param array $requestedattributes
+     * @param fields|null $userfieldsapi as returned by {@see self::get_user_fields_api_mapping()}
+     * @param array|null $userfieldsmapping as returned by {@see self::get_user_fields_api_mapping()}
+     * @return array
+     */
+    protected function get_group_attributes(int $groupid, array $requestedattributes, ?\core_user\fields $userfieldsapi,
+                                            ?array $userfieldsmapping): array {
+        global $DB;
+        $attributes = [];
+
+        if (in_array('group_id', $requestedattributes)) {
+            $attributes['group_id'] = $groupid;
+        }
+
+        if (in_array('group_name', $requestedattributes)) {
+            $attributes['group_name'] = $DB->get_field('groups', 'name', ['id' => $groupid]);
+        }
+
+        // Get group members and their attributes.
+        if ($userfieldsapi) {
+            $fieldssql = $userfieldsapi->get_sql('u', true);
+            $params = $fieldssql->params;
+            $params['groupid'] = $groupid;
+            $data = $DB->get_records_sql("SELECT u.id {$fieldssql->selects}
+                                         FROM {groups_members} gm
+                                         JOIN {user} u ON (u.id = gm.userid)
+                                         {$fieldssql->joins}
+                                         WHERE gm.groupid = :groupid", $params);
+
+            $attributes['members'] = [];
+            foreach ($data as $userdata) {
+                $attributes['members'][] = $this->extract_user_attributes($userdata, $userfieldsmapping);
+            }
+        }
+
+        return $attributes;
+    }
+
+    /**
+     * Create a bridge for the given attempt.
+     *
+     * We already provide our own bridges for the core_question_preview and mod_quiz components.
+     * Otherwise, we try to load a bridge class from the component namespace.
+     *
+     * @param \question_attempt $attempt
+     * @return self
+     */
+    public static function create(\question_attempt $attempt): self {
+        global $DB;
+
+        $usage = $DB->get_record('question_usages', ['id' => $attempt->get_usage_id()], '*', MUST_EXIST);
```

Comment:
Mir ist gerade beim Implementieren von #204 folgendes aufgefallen:

Wir wollen schon beim Aufruf von `/attempt/start` die geforderten Attribute zurückgeben. Ein Aufruf von `get_bridge()` in `start_attempt()` führt allerdings hier zur `dmlreadexception`. Der Attempt wurde noch nicht in der Datenbank gespeichert, weshalb `$attempt->get_usage_id()` hier einen zufälligen String zurückgibt ([siehe](https://github.com/moodle/moodle/blob/0582c77a33fc1e272e621ebbf310a3a1ad1e12fb/public/question/engine/questionusage.php#L53-L58)).

Reply:
Oh nein, das ist ja richtig blöd. Mir fällt dazu auch keine Alternative ein, wie wir sonst noch an die Daten rankommen sollen. Dir?
Dann sollten wir als Implementierungsdetail von qtype_questionpy im Falle von `/attempt/start` eben gar keine Attribute übermitteln. Ich sag ja eh, dass diese Attribute rein optional sind und das LMS alles Mögliche oder nur einen Teil zurückliefern darf.

Können wir überhaupt zuverlässig rausfinden, ob `get_usage_id` eine gültige ID zurückgibt bzw. ob der Attempt nun in der DB gespeichert wurde? Wir könnten überprüfen, ob es sich um einen richtigen Integer bzw. String mit [0-9] handelt, aber durch Zufall könnte das ja trotzdem zutreffen.